### ISSUE TITLE
Strapi admin panel not using host information on fetch requests [4.8.x]

### ISSUE THREAD

**Issue by MRDRMUFN** [strapi/strapi#16215]
Cannot get the initial setup page to load when building from Strapi 4.8.2. However, all configurations and environments being equal changing the version to 4.7.1 allows it to succeed.  From what I can tell in the console the browser is attempting to make an api request to localhost:1337 for /project-type which fails as this is a production site and not locally on the machine. 

Site runs behind nginx proxy to docker container running node:16

The issue im experiencing is very similar to #12624 though it only comes up on the newest version of strapi. 

![Screenshot 2023-03-26 135822](https://user-images.githubusercontent.com/544022/227799816-6d5b711f-ca3f-418b-8573-f9dfd229acfd.png)

![Screenshot 2023-03-26 135858](https://user-images.githubusercontent.com/544022/227799819-b07097af-c985-4d09-80d9-98fe85142e88.png)

![Screenshot 2023-03-26 135919](https://user-images.githubusercontent.com/544022/227799818-18fbd3c1-f380-47d1-9b1b-7c1a19a5f586.png)

![Screenshot 2023-03-26 140140](https://user-images.githubusercontent.com/544022/227799817-1b5eafa4-b077-435b-982c-b6022776f0fc.png)

**Comment by CodeStarrk** [comment]
I have the exact same problem now, nginx, docker etc., pls fix *puppy eyes*

**Comment by b3nab** [comment]
Yes I confirm I have the exactly same problem. I am trying to deploy a production version since 2 weeks without success at all. I tried multiple things but there is still something I am not aware of. Reading the documentation doesn't help unfortunately.

Someone could provide to us poor developers some absolute instructions on what we need to do and WHY is that way so then we could understand and adapt the instructions to our environments?

Honestly I'm getting frustrated by this issue.

**Comment by derrickmehaffy** [comment]
Alright first up, this isn't a bug but I'll do my level best to explain this as detailed as I can.

First off how you should handle this / what we intend:

There is a config file: `./config/server.js` where you set the `url` key, this key SHOULD be absolute eg `api.example.com`. And if you are using an ENV var then this var NEEDS to exist during build time eg `yarn build`.

Now as to why, by default we prebuild the admin panel using localhost:1337 and that is shipped with the node modules, meaning if the server URL isn't set and the admin built then the react admin panel will try to call the Strapi backend over the default localhost:1337. We do this to make running Strapi faster locally for development so that users don't need to build for local dev.

No this URL cannot be dynamic (at least not yet) and no you should not use relative URLs (it can work but in very specific infrastructural configurations). Yes you need to build the admin panel if this URL is different meaning that no you cannot use a single docker container for all environments if you intend for that docker container to serve both the admin and the backend. React is a CSR application meaning the users browser will download all of the files and execute them on their browser which means all API calls will need to know where to go, this means the absolute URL to the Strapi backend needs to exist so that the users browsers knows where to actually send those REST API calls to.

---

TLDR: Configure the URL, make sure you set that URL during build time (this is for you docker users) and build the admin panel.

All of this is documented in the deployment guides for Nginx, HAProxy, and Caddy as it's assumed you are using something for SSL offloading.

- [caddy](https://docs.strapi.io/dev-docs/deployment/caddy-proxy)
- [nginx](https://docs.strapi.io/dev-docs/deployment/nginx-proxy)
- [haproxy](https://docs.strapi.io/dev-docs/deployment/haproxy-proxy)

Sample server.js:

```js
module.exports = ({ env }) => ({
  // ...
  url: env('PUBLIC_URL','https://api.example.com'),
  // ...
});
```

**Comment by derrickmehaffy** [comment]
PS, yes it has been request to be able to change the URL after build, you can find the feature request for that here: https://feedback.strapi.io/developer-experience/p/configure-strapi-backend-url-after-build

No we don't have a timeline for it

**Comment by derrickmehaffy** [comment]
If you are a docker user, I strongly suggest you check out the community dockerize tool which can construct a proper dockerfile where the env vars are handled properly: https://www.npmjs.com/package/@strapi-community/dockerize
